Source/Core/VideoCommon/VertexShaderGen.cpp

```diff
@@ -91,37 +94,135 @@ ShaderCode GenerateVertexShaderCode(APIType api_type, const ShaderHostConfig& ho
   out.Write("{}", s_shader_uniforms);
   out.Write("}};\n");
 
+  if (uid_data->vs_expand != VSExpand::None)
+  {
+    out.Write("UBO_BINDING(std140, 3) uniform GSBlock {{\n");
+    out.Write("{}", s_geometry_shader_uniforms);
+    out.Write("}};\n");
+
+    if (api_type == APIType::D3D)
+    {
+      // D3D doesn't include the base vertex in SV_VertexID
+      out.Write("UBO_BINDING(std140, 4) uniform DX_Constants {{\n"
+                "  uint base_vertex;\n"
+                "}};\n\n");
+    }
+  }
+
   out.Write("struct VS_OUTPUT {{\n");
   GenerateVSOutputMembers(out, api_type, uid_data->numTexGens, host_config, "",
                           ShaderStage::Vertex);
   out.Write("}};\n\n");
 
   WriteIsNanHeader(out, api_type);
 
-  out.Write("ATTRIBUTE_LOCATION({}) in float4 rawpos;\n", SHADER_POSITION_ATTRIB);
-  if ((uid_data->components & VB_HAS_POSMTXIDX) != 0)
-    out.Write("ATTRIBUTE_LOCATION({}) in uint4 posmtx;\n", SHADER_POSMTX_ATTRIB);
-  if ((uid_data->components & VB_HAS_NORMAL) != 0)
-    out.Write("ATTRIBUTE_LOCATION({}) in float3 rawnormal;\n", SHADER_NORMAL_ATTRIB);
-  if ((uid_data->components & VB_HAS_TANGENT) != 0)
-    out.Write("ATTRIBUTE_LOCATION({}) in float3 rawtangent;\n", SHADER_TANGENT_ATTRIB);
-  if ((uid_data->components & VB_HAS_BINORMAL) != 0)
-    out.Write("ATTRIBUTE_LOCATION({}) in float3 rawbinormal;\n", SHADER_BINORMAL_ATTRIB);
-
-  if ((uid_data->components & VB_HAS_COL0) != 0)
-    out.Write("ATTRIBUTE_LOCATION({}) in float4 rawcolor0;\n", SHADER_COLOR0_ATTRIB);
-  if ((uid_data->components & VB_HAS_COL1) != 0)
-    out.Write("ATTRIBUTE_LOCATION({}) in float4 rawcolor1;\n", SHADER_COLOR1_ATTRIB);
-
-  for (u32 i = 0; i < 8; ++i)
+  if (uid_data->vs_expand == VSExpand::None)
```

Comment:
Would it be better to check `host_config.backend_dynamic_vertex_loader`, since that more directly explains what the difference in behavior is, even if we don't normally use the dynamic vertex loader with specialized shaders? I assume (but haven't checked) that `host_config.backend_dynamic_vertex_loader` is implicitly set to true if vs expansion is used.

Reply:
This isn't actually the case, because I leave `backend_dynamic_vertex_loader` false on APIs that didn't previously have issues with ubershader compilation stutter (OpenGL and DX11), even if they support VS expand.
This allows them to default to dynamic vertex loader disabled, but advertise support for VS expand, which will then enable dynamic vertex loader if the setting to use VS expand is enabled.
Do you think I should rename it to `backend_prefers_dynamic_vertex_loader`?  Or rename the option in `VideoConfig`, and set `backend_dynamic_vertex_loader` to `g_ActiveConfig.backend_info.bPrefersDynamicVertexLoader || g_ActiveConfig.UseVSForLinePointExpand()`?

Also, in this specific case, `backend_dynamic_vertex_loader` only enables expansion for ubershaders.  For non-uber shaders, we only need it for shaders that actually expand things.

Reply:
> Also, in this specific case, backend_dynamic_vertex_loader only enables expansion for ubershaders. For non-uber shaders, we only need it for shaders that actually expand things.

From this, it's probably fine to leave the existing behavior alone. I don't really have an opinion on changing `backend_dynamic_vertex_loader` or not at this point.